addons/product/models/product.py

```diff
@@ -880,3 +880,25 @@ def _check_product_variant(self):
         for supplier in self:
             if supplier.product_id and supplier.product_tmpl_id and supplier.product_id.product_tmpl_id != supplier.product_tmpl_id:
                 raise ValidationError(_('The product variant must be a variant of the product template.'))
+
+    @api.onchange('product_tmpl_id')
+    def _onchange_product_tmpl_id(self):
+        """Clear product variant if it no longer matches the product template."""
+        if self.product_id and self.product_id not in self.product_tmpl_id.product_variant_ids:
+            self.product_id = False
+
+    def _sanitize_vals(self, vals):
+        """Sanitize vals to sync product variant & template on read/write."""
+        if 'product_id' in vals and 'product_tmpl_id' not in vals:
```

Comment:
Hi @lvsz 

Instead of `if 'product_id' in vals` we should check `if vals.get('product_id', False)`

Issue: 
Tested on latest build: https://57755695-15-0-all.runbot180.odoo.com/web?debug=1#id=134&cids=1&menu_id=853&action=273&model=product.product&view_type=form

* Go to Product variants
* Add a new pricelist line and save
* Old line is removed

This happens as we are changing `product_tmpl_id` to False as original line doesn't have `product_id`

Do you want me to create odoo ticket/or PR with this fix ? or It will be handled by you/team

Reply:
Oops, didn't even realize you could add vendor pricelists from there. I'll create new PR and link it to the same ticket. Thanks for finding the cause & notifying me.

Reply:
https://github.com/odoo/odoo/pull/152881